cert/cert-renewal/src/cert_interface.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+#[async_trait::async_trait]
+pub trait CertInterface {
+    /// Retrieve a certificate from the provided `cert_id`.
+    async fn get_cert(&mut self, cert_id: &str) -> Result<openssl::x509::X509, crate::Error>;
+
+    /// Retrieve a private key from the provided `key_id`.
+    async fn get_key(
+        &mut self,
+        key_id: &str,
+    ) -> Result<openssl::pkey::PKey<openssl::pkey::Private>, crate::Error>;
+
+    /// Renew the provided certificate.
+    ///
+    /// This function should renew `old_cert` and its key and return the renewed certificate and
+    /// key. It MUST leave `old_cert` and its key intact upon returning; i.e. it must not erase
+    /// `old_cert` or its key.
+    ///
+    /// After this function returns, the renewal engine needs to perform additional checks and
+    /// calculations on the renewed certificate. If the renewal engine determines the new certificate
+    /// to be invalid, it will discard the renewed certificate and fall back to the old credentials.
+    ///
+    /// Once the renewal engine determines the renewed certificate to be valid, it will call
+    /// `write_credentials`. The old certificate and its key may then be overwritten.
+    async fn renew_cert(
+        &mut self,
+        old_cert: &openssl::x509::X509,
+        key_id: &str,
+    ) -> Result<
+        (
+            openssl::x509::X509,
+            openssl::pkey::PKey<openssl::pkey::Private>,
```

Comment:
@arsing The new key to be temporarily stored in memory, as the old key cannot be erased from keyd until the renewal is complete. Thoughts?

Reply:
This problem is only because we decided to rotate the key too. If we use the same key between the old and new certs, we won't have this problem.

Reply:
Resolved: we will rotate the keys and add a way to rename keys in keyd.